### PR TITLE
Collect and send SMART data to graphite

### PR DESCRIPTION
**Pull request by jayofdoom**

### CODE REVIEW

File: onmetal_ironic_hardware_manager/__init__.py [hunk 1]
```diff
@@ -209,8 +215,64 @@ def _list_lsi_devices(self):
...
+
+    def get_disk_metrics(self):
+        smart_data_columns = ['VALUE', 'WORST', 'THRESH', 'RAW_VALUE']
+        block_devices = self.list_block_devices()
+        for block_device in block_devices:
+            if self._is_warpdrive(block_device):
+                # TODO(JayF): Add support for Warpdrive
+                LOG.info('Block device {0} not yet supported for collecting'
+                         'metrics'.format(block_device.name))
+                next
```
Comment (lvh) [inline onmetal_ironic_hardware_manager/__init__.py]: This says `next`, ITYM `continue`.

Comment (lvh) [inline onmetal_ironic_hardware_manager/__init__.py]: However, it doesn't matter, because you have two mutually exclusive branches, so I think it should just be removed.

File: onmetal_ironic_hardware_manager/__init__.py [hunk 2]
```diff
@@ -209,8 +215,64 @@ def _list_lsi_devices(self):
...
+            else:
+                disk_metrics = self._get_smartctl_attributes(block_device)
+                prefix = 'smartdata_{0}'.format(block_device.name)
+                metrics_to_send = {}
+                for k, v in disk_metrics.iteritems():
+                    for heading in smart_data_columns:
+                        key = k + '.' + heading
+                        metrics_to_send[key] = v[heading]
+                self._send_gauges(prefix, metrics_to_send)
+                next
```
Comment (lvh) [inline onmetal_ironic_hardware_manager/__init__.py]: same comment as above about `next` vs `continue`, and also that it probably should just be removed

File: onmetal_ironic_hardware_manager/__init__.py [hunk 3]
```diff
@@ -209,8 +215,64 @@ def _list_lsi_devices(self):
...
+                         'metrics'.format(block_device.name))
+                next
+            else:
+                disk_metrics = self._get_smartctl_attributes(block_device)
+                prefix = 'smartdata_{0}'.format(block_device.name)
+                metrics_to_send = {}
+                for k, v in disk_metrics.iteritems():
+                    for heading in smart_data_columns:
+                        key = k + '.' + heading
+                        metrics_to_send[key] = v[heading]
```
Comment (lvh) [inline onmetal_ironic_hardware_manager/__init__.py]: does this coerce to int anywhere?

File: onmetal_ironic_hardware_manager/__init__.py [hunk 4]
```diff
@@ -105,6 +105,12 @@ def get_decommission_steps(self):
                 'reboot_requested': False,
             },
             {
+                'state': 'get_disk_metrics',
+                'function': 'get_disk_metrics',
+                'priority': 41,
+                'reboot_requested': False
+            },
```
Comment (jimrollenhagen) [inline onmetal_ironic_hardware_manager/__init__.py]: Can I just say here, that I love that we can add new steps without touching Ironic?

File: onmetal_ironic_hardware_manager/__init__.py [hunk 5]
```diff
@@ -209,8 +215,64 @@ def _list_lsi_devices(self):
...
+            if line.strip().startswith('ID#'):
+                header = line.strip().split()
+                break
+
+        attributes = {}
+        for line in it:
+            line = line.strip()
+            if not line:
+                continue
+            linelist = line.split(None)
```
Comment (jimrollenhagen) [inline onmetal_ironic_hardware_manager/__init__.py]: `split(None)` is equivalent to `split()`, fwiw.

File: onmetal_ironic_hardware_manager/__init__.py [hunk 6]
```diff
@@ -209,8 +215,64 @@ def _list_lsi_devices(self):
...
+
+    def _send_gauges(self, prefix, metrics_to_send):
+        """Batch-send gauges to MetricsLogger.
+
+        :param prefix: The prefix given to MetricLogger
+        :param metrics: Dict in the format {'key': 'value'} where key is the
+                        metric name and value is the metric.
+        """
+        logger = metrics.getLogger(prefix)
+        for name, gauge in metrics_to_send.iteritems():
```
Comment (jimrollenhagen) [inline onmetal_ironic_hardware_manager/__init__.py]: `iteritems` isn't python 3 compatible, you could use `six.iteritems` to be compatible with both and have the same behavior.

File: onmetal_ironic_hardware_manager/__init__.py [hunk 7]
```diff
@@ -209,8 +215,64 @@ def _list_lsi_devices(self):
...
+        block_devices = self.list_block_devices()
+        for block_device in block_devices:
+            if self._is_warpdrive(block_device):
+                # TODO(JayF): Add support for Warpdrive
+                LOG.info('Block device {0} not yet supported for collecting'
+                         'metrics'.format(block_device.name))
+                next
+            else:
+                disk_metrics = self._get_smartctl_attributes(block_device)
+                prefix = 'smartdata_{0}'.format(block_device.name)
```
Comment (jimrollenhagen) [inline onmetal_ironic_hardware_manager/__init__.py]: Theoretically it's possible for drives to have different names on different boots. If this happens and our metric key depends on the name (e.g. `/dev/sda`), could get weird. Not sure if there's something better to do here, serial number?

File: onmetal_ironic_hardware_manager/tests/manager.py [hunk 8]
```diff
@@ -111,7 +114,7 @@ def test_erase_block_device_lsi_notfound(self,
                           self.hardware.erase_block_device,
                           self.block_device)
 
-        mocked_execute.assert_has_calls([])
+        mocked_execute.assert_has_calls([], any_order=True)
```
Comment (jimrollenhagen) [inline onmetal_ironic_hardware_manager/tests/manager.py]: As we discussed in IRC, I don't think this is right, probably want `self.assertEqual(0, mocked_execute.call_count)`

File: onmetal_ironic_hardware_manager/tests/manager.py [hunk 9]
```diff
@@ -242,6 +245,268 @@ def test_remove_bootloader(self, mocked_execute):
             'count=1',
             check_exit_code=[0])])
 
+    @mock.patch.object(utils, 'execute')
+    def test__get_smartctl_attributes(self, mocked_execute):
+        expected = {
+ '1-Raw_Read_Error_Rate': {'FLAG': '0x000a',
```
Comment (jimrollenhagen) [inline onmetal_ironic_hardware_manager/tests/manager.py]: I'd love lines like this to be indented properly. :)

File: onmetal_ironic_hardware_manager/tests/manager.py [hunk 10]
```diff
@@ -242,6 +245,268 @@ def test_remove_bootloader(self, mocked_execute):
...
+                      'VALUE': '100',
+                      'WHEN_FAILED': '-',
+                      'WORST': '100'}}
+
+        mocked_execute.return_value = SMARTCTL_ATTRIBUTES_OUT
+        self.block_device = hardware.BlockDevice('/dev/sda', '32G MLC SATADOM',
+                                                 31016853504, False)
+        actual = self.hardware._get_smartctl_attributes(self.block_device)
+
+        mocked_execute.assert_has_calls([mock.call(
```
Comment (jimrollenhagen) [inline onmetal_ironic_hardware_manager/tests/manager.py]: Again, talked about this earlier, should check against `mocked_execute.call_args_list`.

File: onmetal_ironic_hardware_manager/tests/manager.py [hunk 11]
```diff
@@ -242,6 +245,268 @@ def test_remove_bootloader(self, mocked_execute):
...
+    def test_get_disk_metrics(self, mocked_execute):
+        mocked_execute.return_value = SMARTCTL_ATTRIBUTES_OUT
+        self.hardware._send_gauges = mock.Mock()
+        self.hardware.list_block_devices = mock.Mock()
+        self.hardware.list_block_devices.return_value = [hardware.BlockDevice(
+            '/dev/sda', '32G MLC SATADOM', 31016853504, False)]
+
+        self.hardware.get_disk_metrics()
+
+        mocked_execute.assert_has_calls([mock.call(
```
Comment (jimrollenhagen) [inline onmetal_ironic_hardware_manager/tests/manager.py]: Ditto

**Comment by jimrollenhagen** [comment]
Looks good overall, some comments inline :)

File: onmetal_ironic_hardware_manager/__init__.py (hunk 3 above)
Comment (jayofdoom) [inline onmetal_ironic_hardware_manager/__init__.py]: No, and I don't think graphite will consider any differences in int vs string in metrics.

File: onmetal_ironic_hardware_manager/__init__.py [hunk 12]
```diff
@@ -209,8 +217,66 @@ def _list_lsi_devices(self):
...
+                prefix = 'smartdata_{0}_{1}'.format(
+                        os.path.basename(block_device.name),
+                        block_device.model.translate(None, " "))
+                metrics_to_send = {}
+                for k, v in six.iteritems(disk_metrics):
+                    if v['RAW_VALUE'] == '0':
+                        continue
+                    for heading in smart_data_columns:
+                        key = k + '.' + heading
+                        metrics_to_send[key] = v[heading]
```
Comment (aweeks) [inline onmetal_ironic_hardware_manager/__init__.py]: Not really a big deal right now, but you're sending the metric values as strings.  This works coincidentally, because the metrics library currently uses the str(value) to send gauges, but in future versions, it may perhaps only accept numeric values.

File: onmetal_ironic_hardware_manager/__init__.py [hunk 13]
```diff
@@ -209,46 +217,161 @@ def _list_lsi_devices(self):
...
+            for line in it:
+                line = line.strip()
+                if not line:
+                    continue
+                if line.startswith('Data Slot #'):
+                    drivenum = line.split()[3][0]
+                    driveserial = line.split()[7]
+                    attrkey = drivenum + '_' + driveserial
+                    attributes[attrkey] = {}
+                if line.startswith('-------------- Cumulative'):
```
Comment (pcsforeducation) [inline onmetal_ironic_hardware_manager/__init__.py]: I think you can kill this if and combine the two loops.

**Comment by pcsforeducation** [comment]
Overall looking good. Could use some comments in the text processing portions, or this is going to be really hard to parse in a month without comparing all the lines to the test data.

File: onmetal_ironic_hardware_manager/__init__.py (hunk 13 above)
Comment (jayofdoom) [inline onmetal_ironic_hardware_manager/__init__.py]: Not at all. Basically what I'm doing here is "fast forwarding" from the "current" stats (which is only for this power cycle, and is not interesting) to the part with the cumulative stats.

File: onmetal_ironic_hardware_manager/__init__.py (hunk 12 above)
Comment (jayofdoom) [inline onmetal_ironic_hardware_manager/__init__.py]: I can force these to int or double if you want, but if this will work for now I'd rather leave that for later.

File: onmetal_ironic_hardware_manager/__init__.py [hunk 14]
```diff
@@ -209,46 +217,181 @@ def _list_lsi_devices(self):
...
+                line = line.strip()
+                # note(JayF): It doesn't make sense to store time in graphite
+                if not line or line.endswith('(Hours:Minutes:Seconds)'):
+                    continue
+                # note(JayF): This is the first line of the footer. If we get
+                # here, we're done parsing.
+                if line.startswith('Warranty Remaining'):
+                    break
+                # note(JayF): Some of the metrics have units or notes in
+                # parens, we look for these and adjust our rsplit accordingly
```
Comment (jimrollenhagen) [inline onmetal_ironic_hardware_manager/__init__.py]: Hmm, since we're not grabbing index 2 or 3 at all, do we actually need to adjust the number of splits?

File: onmetal_ironic_hardware_manager/__init__.py (hunk 12 above)
Comment (jimrollenhagen) [inline onmetal_ironic_hardware_manager/__init__.py]: +1, deal with it later.

**Comment by jimrollenhagen** [comment]
+1, this LGTM, one thing inline that may be able to be cleaned up a bit.

Would love other eyes on this as well.

File: onmetal_ironic_hardware_manager/__init__.py (hunk 14 above)
Comment (jayofdoom) [inline onmetal_ironic_hardware_manager/__init__.py]: Yes, because otherwise the line is split incorrectly. Note I'm splitting from the right and taking the two leftmost items.

Comment (jimrollenhagen) [inline onmetal_ironic_hardware_manager/__init__.py]: mmm, rsplit. right. carry on :)

File: onmetal_ironic_hardware_manager/__init__.py (hunk 13 above)
Comment (pcsforeducation) [inline onmetal_ironic_hardware_manager/__init__.py]: Ah! Missed that iter() line.

**Comment by pcsforeducation** [comment]
+1 Thanks for the comments!

File: onmetal_ironic_hardware_manager/__init__.py [hunk 15]
```diff
@@ -209,46 +217,181 @@ def _list_lsi_devices(self):
...
+
     def _erase_lsi_warpdrive(self, block_device):
-        if block_device.model != LSI_MODEL:
+        if not self._is_warpdrive(block_device):
             return False
 
         # NOTE(JayF): Start timing here, so any devices short circuited
         # don't produce invalidly-short metrics.
         with metrics.instrument_context(__name__, 'erase_lsi_warpdrive'):
-            device_name = os.path.basename(block_device.name)
```
Comment (supermari0) [inline onmetal_ironic_hardware_manager/__init__.py]: this is excellent - makes this much more readable

**Comment by supermari0** [comment]
LGTM

**Comment by supermari0** [comment]
Merging.
